PICARD-1273: Add an option to exclude new cover art type "Raw / Unedited"

rdswift opened the pull request:
<!--
    Hello! Thanks for submitting a pull request to MusicBrainz Picard. We
    appreciate your time and interest in helping our project!

    Use this template to help us review your change. Not everything is required,
    depending on your change. Keep or delete what is relevant for your change.
    Remember that it helps us review if you give more helpful info for us to
    understand your change.

    Ensure that you've read through and followed the Contributing Guidelines, in
    [CONTRIBUTING.md](https://github.com/metabrainz/picard/blob/master/CONTRIBUTING.md).
-->

# Summary

<!--
    Update the checkbox with an [x] for the type of contribution you are making.
-->

* This is a…
    * [ ] Bug fix
    * [x] Feature addition
    * [ ] Refactoring
    * [ ] Minor / simple change (like a typo)
    * [ ] Other
* **Describe this change in 1-2 sentences**: This change provides the user with the option of excluding selected cover art types from downloading from CAA.

# Problem

<!--
    Anything that helps us understand why you are making this change goes here.
    What problem are you trying to fix? What does this change address?
-->
[STYLE-980](https://tickets.metabrainz.org/browse/STYLE-980) has added a new cover art style "Raw / Unedited".  Although beneficial for identifying and allowing better images for archival purposes, the images within this type are generally not suitable for tagging purposes.  Picard currently allows the user to select the types of images to download from CAA, but does not provide the ability to exclude selected image types from being downloaded.

* JIRA ticket: [PICARD-1273](https://tickets.metabrainz.org/browse/PICARD-1273): Add an option to exclude new cover art type "Raw / Unedited"
<!--
    Please make sure you prefix your pull request title with 'PICARD-XXX' in order
    for our ticket tracker to link your pull request to the relevant ticket.
-->



# Solution

<!--
    The details of your change. Talk about technical details, considerations, or
    other interesting points. If you have a lot to say, be more detailed in this
    section.
-->
This change adds new user settings to indicate whether selected image types should be excluded from downloads from CAA and the list of image types to be excluded, similar to the option to only include selected image types.  Additional logic is applied to the list of images selected for download, and any images with a type intersecting the user's "exclusion" type list are removed from the download list.

Generally speaking, the current logic for determining whether to download an image is (where the image is downloaded when "types" evaluates to True):

    if self.restrict_types:
        # only keep enabled caa types
        types = set(image["types"]).intersection(
            set(self.caa_types))
    else:
        types = True

This change adds a further check for exclusion, as:

    if self.restrict_types:
        # only keep enabled caa types
        types = set(image["types"]).intersection(
            set(self.caa_types))
    else:
        types = True
    if types and self.omit_types:
        # omit selected caa types
        types = not set(image["types"]).intersection(
            set(self.caa_types_to_omit))



# Action

<!--
    Other than merging your change, do you want / need us to do anything else
    with your change? This could include reviewing a specific part of your PR.
-->

Currently, it appears that CAA does not return the new "Raw/Unedited" type in the list of image types associated with an image.  This is described (with an example) in [CAA-113](https://tickets.metabrainz.org/browse/CAA-113): CAA does not return "Raw/Unedited" in image type array.  This will need to be addressed before the proposed changes will work.

samj1912 commented:
I am not sure if creating another dialog is the right idea.

Let's take this for example - 

Current usage allows a user to select AND exclude a particular type of cover art.

Maybe fix the logic there or unify the dialogs somehow.

rdswift commented:
@samj1912, I can see how it could be confusing for the user.  What do you think about just having one selection list with the choices of "Ignore", "Include" and "Exclude" for each cover art image type?  The choices could be done using radio buttons or a combo box (my preference) for each image type.

rdswift commented:
@samj1912 Here is an example of the combo box option.  Is this more what you had in mind?

![image](https://user-images.githubusercontent.com/16990904/44006048-93a7c2f8-9e3a-11e8-9cb3-a9904dd778ce.png)

rdswift commented:
This is another option using radio buttons.  I actually think I prefer this one.  Comments?

![image](https://user-images.githubusercontent.com/16990904/44008737-e14441d4-9e63-11e8-93e6-6ac9d379e683.png)

zas commented:
I tend to prefer the combo box option. The radio buttons one looks overloaded.

rdswift commented:
Thanks, @zas.  I'll proceed on that basis and update the PR today or tomorrow.  Are the terms "Include", "Exclude" and "Ignore" okay, or can you suggest something more appropriate?

rdswift commented:
@zas / @samj1912 

I've updated the code to use the combo box option for selecting which file types to include or exclude.  That simplified the code significantly, with only one file being modified.

For testing, I have been using https://musicbrainz.org/release/a4a99919-0923-40aa-8f64-b17820ebc407/cover-art

rdswift commented on picard/coverart/providers/caa.py:
```diff
@@ -82,15 +91,30 @@ def __init__(self, parent=None, types=None):
+            elif name in types_exclude:
+                item.setCurrentIndex(2) # Exclude type
+            else:
+                item.setCurrentIndex(1) # Ignore type
```
Changed to use constants defined as per your earlier comment.

rdswift commented on picard/coverart/providers/caa.py:
```diff
@@ -69,11 +69,20 @@ def caa_url_fallback_list(desired_size, thumbnails):
+    # Text displayed in combo box options.
+    _include = N_('Include')
+    _exclude = N_('Exclude')
+    _ignore = N_('Ignore')
```
Dropped these entirely and moved them to the `addItems()` method call as `item.addItems([ _('Include'), _('Ignore'), _('Exclude'),])`.

rdswift commented on picard/coverart/providers/caa.py:
```diff
@@ -320,6 +357,9 @@ def _caa_json_downloaded(self, data, http, error):
             else:
+                if self.restrict_types:
+                    log.debug('Include types: %s' % (self.caa_types,))
+                    log.debug('Exclude types: %s' % (self.caa_types_to_omit,))
```
Changed as suggested.  Also added clarification to the accepted / rejected debug log entry.

rdswift commented on picard/coverart/providers/caa.py:
```diff
@@ -82,15 +91,30 @@ def __init__(self, parent=None, types=None):
-            text = translate_caa_type(name)
-            item = QtWidgets.QCheckBox(text)
-            item.setChecked(name in types)
+            text = translate_caa_type(name) + ':'
```
I dropped the colon and revised the code to remove the `text` variable.  You're right, in that it served no real purpose.  Moved the `translate_caa_type(name)` directly to the argument of the `item.setText()` method call.

rdswift commented on picard/coverart/providers/caa.py:
```diff
@@ -126,26 +151,35 @@ def help(self):
+    def get_selected_types_exclude(self):
         return [typ['name'] for item, typ in self._items.items() if
-                item.isChecked()] or ['front']
+                item.currentIndex() == 2] or ['none']
```
Changed to use explicit constants as per your previous comment.

phw commented:
I think the average user will struggle understanding the difference between "ignore" and "exclude", and to be honest I am not 100% sure I got it myself. Can we come up with a better naming?

The old dialog with only the checkboxes was ok to understand: Just choose the cover art types you are interested in.

As I understand it the new "exclude" option will blacklist certain types, right? So if I have "Front" included, "Raw/Unedited" excluded and all other types ignored it will use images of e.g. type "Front", or "Front, Back", but never anything that has "Raw/Unedited"?

I think the main confusion will arise because people probably don't intuitively think about cover art having multiple types.  And "Ignore" sounds a lot like as if images of this type will not be used. So I think the dialog should explain a bit more.

One simple solution would be to just add a small help section with some info like:

> Include: Images having this type set will be downloaded and used as cover art
> Exclude: Images having this type will never be downloaded, even if they have also an included type
> Ignore: This type has no affect on choosing the cover art

I struggle with describing "Ignore", maybe somebody has a better idea.

Another option would be to rethink the UI. Maybe something like this?

![grafik](https://user-images.githubusercontent.com/29852/44166524-db215480-a0cb-11e8-9304-dd0ab8d7d14d.png)

Not sure really how to best solve this, but I am very sure Include, Ignore, Exclude will be confusing for many.

Apart from that I think this is really useful functionality. Not only for the raw type but also for other cases. E.g. for my files I basically never want to use an image labeled as front that also includes the back.

rdswift commented:
I agree that the terms could very well be confusing for most users, at least initially.  I have struggled all along with trying to find the right terms to use.  Your simple solution with the help section explanation could work.  Perhaps the explanation for "Ignore" could be something like, "This type is not considered when deciding whether or not to download an image."  Perhaps "Ignore" should be replaced with "Not considered"?

The intended logic is simply that an image will be downloaded only if a) one of its types matches a type in the "Include" list, **AND** b) none of its types match a type in the "Exclude" list.

I like your alternate UI suggestion because I think that it makes it much clearer.  Unfortunately, the way I struggle with UI design using Qt, I suspect it will take me quite a while to develop tight working code for it.  :frowning_face:

rdswift commented:
Does something like this make it a bit clearer for the users?

![image](https://user-images.githubusercontent.com/16990904/44228034-71e42480-a151-11e8-885f-888f310757fe.png)

samj1912 commented:
I like the solution proposed by phw and the one implemented by you in the 2nd image.

phw commented:
I just ran the code, that feels much better now! I like it very much.

zas commented:
It looks much better now ! Good job :+1: 

Few quirks with the UI:

![capture du 2018-08-18 21-46-46](https://user-images.githubusercontent.com/151042/44302808-64d15c00-a330-11e8-9181-494e24fc9652.png)

Since one can select one item in each list it becomes hard to understand, imho only one item should be selected at once.

Also items should be ordered consistently, for example, if i move one item from top of middle list to exclude list then back to the middle list, it appears at the end of the list. May be sorting alphabetically the (translated) items would improve that, dunno.

When an action isn't possible, matching button should be disabled (no item selected, or empty list).

I'll test a bit more tomorrow.

zas commented:
@rdswift : thanks :)

Let's nitpick a bit:
- it would be great to be able to select multiple items in the same list (using shift/control) and move them at once.
- instead of Clear All it may be better to have a Reset to Defaults button.
- may be start a thread on forums about this dialog, and see what users think about it.

What do you think?

rdswift commented:
@zas 

Just updated it to allow multiple selections.  I thought it was already doing that, but it seems that I forgot to specifically enable it.

Added a ***Restore Defaults*** button.  Great suggestion!  For now I've also left the ***Clear All*** button, but that can easily be removed if we want.

I'm just working on the thread in the forums now, so we'll see what the community has to say and what suggestions they might have.  Another great idea.  Thanks.

zas commented on picard/coverart/providers/caa.py:
```diff
@@ -46,6 +46,61 @@
+        item = list1.item(0)
+        move_item(item, list1, list2)
+
+def clear_listbox_focus(list_to_clear):
```
This one isn't strictly needed, you can call `clearSelection()` directly.

rdswift replied:
Right you are.  That was a holdover from when I was clearing each item individually.

zas commented on picard/coverart/providers/caa.py:
```diff
@@ -46,6 +46,61 @@
+    list_box.setSizePolicy(QtWidgets.QSizePolicy.Minimum, QtWidgets.QSizePolicy.Expanding)
+    list_box.setSortingEnabled(True)
+    list_box.setSelectionMode(QtWidgets.QAbstractItemView.ExtendedSelection)
+    return list_box
```
Instead of using `make_*()`, you could subclass Qt objects

```python
class ListBox(QtWidgets.QListWidget):
    def __init__(self, parent=None):
        super().__init__(parent=parent)
        self.setFixedSize(QtCore.QSize(150, 250))
        self.setSizePolicy(QtWidgets.QSizePolicy.Minimum, QtWidgets.QSizePolicy.Expanding)
        self.setSortingEnabled(True)
        self.setSelectionMode(QtWidgets.QAbstractItemView.ExtendedSelection)
```

```python
class ArrowButton(QtWidgets.QPushButton):
    def __init__(self, text, parent=None, command=None):
        super().__init__(text, parent=parent)
        if command:
            self.clicked.connect(command)
        self.setFixedSize(QtCore.QSize(35, 20))
```

And same for `make_arrows_column()`

rdswift replied:
Makes sense.  It seems I haven't completely embraced the concept of OOP yet.  Old habits die hard.  :grin:

zas commented on picard/coverart/providers/caa.py:
```diff
@@ -46,6 +46,61 @@
+    """Move the specified item from one listbox to another.  Used in the CAA image type selection dialog."""
+    clone = item.clone()
+    list2.addItem(clone)
+    list1.takeItem(list1.row(item))
```
Perhaps you can skip the `clone()` if you remove it first from list1 then add it to list2:

```python
list2.addItem(list1.takeItem(list1.row(item)))
```

Untested though.

rdswift replied:
It works, but I had to remove it from the source list before I could add it to the target list.

zas commented on picard/coverart/providers/caa.py:
```diff
@@ -123,8 +72,97 @@ def caa_url_fallback_list(desired_size, thumbnails):
+    ARROW_BUTTON_WIDTH = 35
+    ARROW_BUTTON_HEIGHT = 20
+
+    def __init__(self, label, command=None, parent=None ):
```
There's an extra space before last parenthesis ;)

rdswift replied:
And that's something I even look for...  Argh!  :blush:

zas commented on picard/coverart/providers/caa.py:
```diff
@@ -123,8 +72,97 @@ def caa_url_fallback_list(desired_size, thumbnails):
     return urls
 
 
+class arrow_button(QtWidgets.QPushButton):
```
Class names should be camel-cased, so it should be `class ArrowButton`

rdswift replied:
Fixed.

zas commented on picard/coverart/providers/caa.py:
```diff
@@ -123,8 +72,97 @@ def caa_url_fallback_list(desired_size, thumbnails):
+        self.setFixedSize(QtCore.QSize(self.ARROW_BUTTON_WIDTH, self.ARROW_BUTTON_HEIGHT))
+
+
+class arrows_column(QtWidgets.QWidget):
```
`ArrowColumn`

rdswift replied:
Fixed.

zas commented on picard/coverart/providers/caa.py:
```diff
@@ -123,8 +72,97 @@ def caa_url_fallback_list(desired_size, thumbnails):
+    def __init__(self, list_add=None, list_add_all=None, list_remove=None, list_remove_all=None, reverse=False, parent=None):
+        _spacer_item = QtWidgets.QSpacerItem(20, 20, QtWidgets.QSizePolicy.Minimum, QtWidgets.QSizePolicy.Expanding)
+        _arrows_layout = QtWidgets.QVBoxLayout()
+        super().__init__(parent=parent)
```
Call it as soon as possible.

rdswift replied:
Moved to first line of __init__.  Same change made to other new classes.

zas commented on picard/coverart/providers/caa.py:
```diff
@@ -123,8 +72,97 @@ def caa_url_fallback_list(desired_size, thumbnails):
+        _arrows_layout = QtWidgets.QVBoxLayout()
+        super().__init__(parent=parent)
+        _arrows_layout.addItem(QtWidgets.QSpacerItem(_spacer_item))
+        self.BUTTON_ADD = arrow_button('>' if reverse else '<', list_add)
```
Since those are widgets, not constants, names should be lowercased.

rdswift replied:
Fixed.

zas commented on picard/coverart/providers/caa.py:
```diff
@@ -123,8 +72,97 @@ def caa_url_fallback_list(desired_size, thumbnails):
+
+    def __init__(self, list_add=None, list_add_all=None, list_remove=None, list_remove_all=None, reverse=False, parent=None):
+        _spacer_item = QtWidgets.QSpacerItem(20, 20, QtWidgets.QSizePolicy.Minimum, QtWidgets.QSizePolicy.Expanding)
+        _arrows_layout = QtWidgets.QVBoxLayout()
```
Since `_arrows_layout` and `_space_item` are local to this method, no need to prefix with `_` (unlike a "private" property, ie. `self._arrows_layout`)

rdswift replied:
Fixed.  Looks much better.

zas commented on picard/coverart/providers/caa.py:
```diff
@@ -123,8 +72,97 @@ def caa_url_fallback_list(desired_size, thumbnails):
+        self.setLayout(_arrows_layout)
+
+
+class list_box(QtWidgets.QListWidget):
```
`class ListBox`

rdswift replied:
Fixed.

rdswift commented:
I think I've addressed all the style issues and such from your last request.  Sorry, I should have re-read the project guidelines again about style, so you wouldn't have to worry about that.  Thanks for your patience.  I think it's looking much better than it was, and I'm learning a lot in the process.  Thank-you all.

zas commented on picard/coverart/providers/caa.py:
```diff
@@ -68,34 +73,209 @@ def caa_url_fallback_list(desired_size, thumbnails):
+        for caa_type in CAA_TYPES:
+            name = caa_type['name']
+            title = _(caa_type['title'])
+            self.image_types_by_display_title[title] = name
```
Using translated string as key isn't a very good idea imho.
You should use the name which is meant for that (`caa_type['name']`)

Also `_()` will not work with caa types, because they come from another set of po files, you need to use `translate_caa_type()` (from coverart.utils)

Read about http://doc.qt.io/qt-5/qlistwidgetitem.html#setData and UserRole i think it is what you need here:
each item has caa type name as data, and you use `QListWidget.sortItems()`
It will prolly not work as expected because of QLocale stuff though (that's something we need to fix in Picard), don't bother with that for now.

samj1912 replied:
I am sure something like User Role will work. The easiest way is to just subclass it and add a field. See the toolbar customization options dialog in UI options for the relevant code.

rdswift replied:
I struggled with that before but I think I have it now.  I'm storing the `caa_type['name']` in the list item ***UserRole*** role and the `caa_type['title']` as the display text.  I've also changed it to use `translate_caa_type()`.  It seems to be sorting the lists properly here for my testing, but everything is English so it's not much of a test.  I removed that extra dictionary.

This may sound strange, but one of my biggest problems is finding out which PyQt5 module a class or enumeration is in.  I had a heck of a time finding UserRole because I tried specifying it in PyQt5.Qt module instead of the PyQt5.QtCore.Qt module.  I guess I'll figure it out eventually.  :slightly_smiling_face: 

I'm just doing some final testing and clean-up and will upload shortly.

rdswift commented on picard/coverart/providers/caa.py:
```diff
@@ -122,30 +304,79 @@ def __init__(self, parent=None, types=None):
+        has_items_exclude = self.list_exclude.count()
+        has_items_ignore = self.list_ignore.count()
+
+        has_selected_include = True if self.list_include.selectedItems() else False
```
I tried that before without the bool() and got the following error:

    Traceback (most recent call last):
      File ".\picard\coverart\providers\caa.py", line 443, in select_caa_types
        self, config.setting["caa_image_types"], config.setting["caa_image_types_to_omit"])
      File ".\picard\coverart\providers\caa.py", line 380, in run
        dialog = CAATypesSelectorDialog(parent, types_include, types_exclude)
      File ".\picard\coverart\providers\caa.py", line 307, in __init__
        self.fix_buttons()
      File ".\picard\coverart\providers\caa.py", line 363, in fix_buttons
        self.arrows_include.button_add.setEnabled(has_items_ignore and has_selected_ignore)
    TypeError: setEnabled(self, bool): argument 1 has unexpected type 'list'

It seems that the error occurs because it's trying to AND a bool with a list.  That's why I went with forcing it to a True or False directly.

It works fine with the bool(), so I'll make that change.

rdswift commented on picard/coverart/providers/caa.py:
```diff
@@ -122,30 +304,79 @@ def __init__(self, parent=None, types=None):
+        self.fill_lists(_CAA_IMAGE_TYPE_DEFAULT_INCLUDE, _CAA_IMAGE_TYPE_DEFAULT_EXCLUDE)
+        self.fix_buttons()
+
+    def fix_buttons(self):
```
You're right.  It doesn't sound good.  I'll change it to ***set_buttons_enabled_state***.

rdswift replied:
I also just realized that I was no longer calling the method since moving the item moving code to ListBox methods.  I've set it up as a callback now and it's working properly again.

rdswift commented:
For some reason appveyor is stalling at the point of ***running test***, and then timing and failing after an hour.  Any idea what I might have done to cause it to fail?

![image](https://user-images.githubusercontent.com/16990904/44443324-72871b80-a594-11e8-83d6-8a62f66b1290.png)

zas commented:
> Any idea what I might have done to cause it to fail?

I tried to restart the build, etc.. and it still gets stuck at start of `python setup.py test -v` command, i see nothing in your PR that may cause this :(

@samj1912 ^^ any idea why this happens ??

phw commented:
For a quick test I ran the tests for that branch in my Windows 10 VM, but that works as expected. And I don't see anything either that looks suspicious :(

samj1912 commented:
Appveyor complains that this pr is non mergable.

rdswift commented:
@samj1912:  I suspect that's because the master branch had changed with all the includes sorted with isort.  I rebased the PR on the updated master, and it appears that appveyor is stalling at the tests again.

zas commented:
> I rebased the PR on the updated master, and it appears that appveyor is stalling at the tests again

It is unrelated to conflicts, i emailed Appveyor support team about it, since we have no clue about what is happening.

rdswift commented:
Before I do something I shouldn't, would it be okay if I (temporarily) modified the `appveyor.yml` file and tried connecting as described in their response to see if I can get any more information?  I would then revert the changes.

phw commented:
@rdswift What I usually do to fix issues with  automated CI is create a separate branch just for experimentation of the builds. And once it is working I commit the final version to the target branch.

rdswift commented:
@phw: Thanks for the tip.  It worked like a charm.

@zas: The error message dialog box was:

![image](https://user-images.githubusercontent.com/16990904/44621693-f167a700-a867-11e8-92f3-ec1fec874c1b.png)

When I clicked on the `Okay` button, the console showed:

![image](https://user-images.githubusercontent.com/16990904/44621755-1f012000-a869-11e8-9004-ee6acbc6f5b8.png)


I have no idea what in my code might be triggering that, but at least it's a place to start...

zas commented:
> I have no idea what in my code might be triggering that, but at least it's a place to start...

Nor i, i send the image to appveyor support ... let's hope they have an idea about what is going on ;)

zas commented on picard/coverart/providers/caa.py:
```diff
@@ -26,8 +26,10 @@
+from functools import partial
 
 from PyQt5 import (
+    Qt,
```
This is prolly the cause of the appveyor issue, see https://stackoverflow.com/questions/48514027/spyder3-python-ide-does-not-start-this-windows-version-does-not-support-the-re

In fact, it is incorrect, and this import should be removed, because `QListWidgetItem` belongs to `QtWidgets`

So replace:

```python
 item = Qt.QListWidgetItem(title)
```

with:

```python
 item = QtWidgets.QListWidgetItem(title)
```

And let's see if it fixes the issue.

zas commented on picard/coverart/providers/caa.py:
```diff
@@ -94,34 +98,203 @@ def caa_url_fallback_list(desired_size, thumbnails):
+        gridlayout.addWidget(self.arrows_exclude, row, column)
+
+        column = 4
+        gridlayout.addWidget(self.list_exclude, row, column)
```
Code would be perhaps easier to read using a small local function, i was thinking about something like:

```python
        self.arrows_include = ArrowsColumn(
            self.list_include,
            self.list_ignore,
            callback=self.set_buttons_enabled_state,
        )

        self.arrows_exclude = ArrowsColumn(
            self.list_exclude,
            self.list_ignore,
            callback=self.set_buttons_enabled_state,
            reverse=True
        )

        def add_widget(row=0, column=0, widget):
             gridlayout.addWidget(widget, row, column)

        add_widget(row=0, column=0, QtWidgets.QLabel(_("Include types list")))
        add_widget(row=1, column=0, self.list_include)

        add_widget(row=1, column=1, self.arrows_include)
        add_widget(row=1, column=2, self.list_ignore)
        add_widget(row=1, column=3, self.arrows_include)

        add_widget(row=0, column=4, QtWidgets.QLabel(_("Exclude types list")))
        add_widget(row=1, column=4, self.list_exclude)
```

No strong opinion here though.

rdswift replied:
I like it.  It definitely looks cleaner than the way it is currently.

zas commented on picard/coverart/providers/caa.py:
```diff
@@ -148,30 +323,97 @@ def __init__(self, parent=None, types=None):
+        return [self.list_include.item(index).data(QtCore.Qt.UserRole) for index in range(self.list_include.count())] or ['front']
+
+    def get_selected_types_exclude(self):
+        return [self.list_exclude.item(index).data(QtCore.Qt.UserRole) for index in range(self.list_exclude.count())] or ['none']
```
A simple iterator in `ListBox` class could help to shorten lines here:

```python
class ListBox(QtWidgets.QListWidget):

    def all_items_data(self, role=QtCore.Qt.UserRole):
        for index in range(self.count()):
            yield self.item(index).data(role)
```

```python
    def get_selected_types_include(self):
        return [self.list_include.all_items_data()] or ['front']

    def get_selected_types_exclude(self):
        return [self.list_exclude.all_items_data()] or ['none']
```

zas commented on picard/coverart/providers/caa.py:
```diff
@@ -94,34 +98,195 @@ def caa_url_fallback_list(desired_size, thumbnails):
+    ARROW_BUTTON_HEIGHT = 20
+
+    def __init__(self, label, command=None, parent=None):
+        if label is None:
```
Imho you can make `label` mandatory, and remove those 2 lines.

rdswift replied:
As written, those lines didn't make sense because `label` was already mandatory.  I've removed them.
